plugins/maven-dependency-resolver/src/main/java/org/robolectric/internal/dependency/MavenArtifactFetcher.java

```diff
@@ -0,0 +1,198 @@
+import com.google.common.io.Files;
+import com.google.common.util.concurrent.AsyncCallable;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Base64;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import org.robolectric.util.Logger;
+
+/**
+ * Class responsible for fetching artifacts from Maven. This uses a thread pool of size two in order
+ * to parallelize downloads.
+ */
+@SuppressWarnings("UnstableApiUsage")
+public class MavenArtifactFetcher {
+
+  private final String repositoryUrl;
+  private final String repositoryUserName;
+  private final String repositoryPassword;
+  private final File localRepositoryDir;
+  private final ExecutorService executorService;
+  private File stagingRepositoryDir;
+
+  public MavenArtifactFetcher(
+      String repositoryUrl,
+      String repositoryUserName,
+      String repositoryPassword,
+      File localRepositoryDir,
+      ExecutorService executorService) {
+    this.repositoryUrl = repositoryUrl;
+    this.repositoryUserName = repositoryUserName;
+    this.repositoryPassword = repositoryPassword;
+    this.localRepositoryDir = localRepositoryDir;
+    this.executorService = executorService;
+  }
+
+  public void fetchArtifact(MavenJarArtifact artifact) {
+    // Assume that if the file exists in the local repository, it has been fetched successfully.
+    if (new File(localRepositoryDir, artifact.jarPath()).exists()) {
+      Logger.info(String.format("Found %s in local maven repository", artifact));
+      return;
+    }
+    this.stagingRepositoryDir = Files.createTempDir();
+    this.stagingRepositoryDir.deleteOnExit();
+    createArtifactSubdirectory(artifact, stagingRepositoryDir);
+    try {
+      Futures.whenAllSucceed(
+              fetchToStagingRepository(artifact.pomSha1Path()),
+              fetchToStagingRepository(artifact.pomPath()),
+              fetchToStagingRepository(artifact.jarSha1Path()),
+              fetchToStagingRepository(artifact.jarPath()))
+          .callAsync(
+              () -> {
+                // double check that the artifact has not been installed
+                if (new File(localRepositoryDir, artifact.jarPath()).exists()) {
+                  removeArtifactFiles(stagingRepositoryDir, artifact);
+                  return Futures.immediateFuture(null);
+                }
+                createArtifactSubdirectory(artifact, localRepositoryDir);
+                boolean pomValid = validateStagedFiles(artifact.pomPath(), artifact.pomSha1Path());
+                if (!pomValid) {
+                  throw new AssertionError("Unable to validate POM file");
```

Comment:
Nit: You might want to be more specific, like from which artifact the POM (or later JAR) file could not be validated.

Reply:
Done, and thanks for the suggestion!